content/docs/document-lifecycle.md

```diff
@@ -0,0 +1,55 @@
+
+Once all `LangiumDocument`s have been created, the `DocumentBuilder` service will sequentially process each `LangiumDocument` as described below.
+
+## Indexing of the AST
+Initial indexing of the AST is executed on `LangiumDocument`s with the state `Parsed`. The default `AstNodeDescriptionProvider` service creates an `AstNodeDescription` for the root node (i.e. the node created by parsing the entry rule) and each named `AstNode` directly descending from the root node. This `AstNodeDescription` contains the `type` of the node, its identifier (i.e. the `name` property), the uri of the document where the node is located, and the location of the node inside of the document. The generated array of `AstNodeDescription`s is then mapped into a `simpleIndex`. This index makes accessible named nodes from a `LangiumDocument` to other `LangiumDocument`s in the same workspace.
+
+Once the initial indexing is done, the document's state is set to `IndexedContent`.
+
+## Pre-processing
+This step is executed on 'LangiumDocument's with the state `IndexedContent` and regroups all necessary steps that needs to be done **prior to** resolving cross-references.
+
+By default, the pre-processing consists of computing the scope of the AST. The default implementation of the `ScopeComputation` service computes the scope for **every** node in the AST and makes named nodes visible to their container. This means that the container holds information about which named nodes are nested inside of it. This is then used during linking to find the closest `AstNode` that matches a reference.
+
+After all pre-processing steps have been completed, the document's state is set to `Processed`.
+
+Please note that resolution of cross-references are not permitted until this point.
+
+## Linking
+Once all pre-processing steps are complete, cross-references are resolved via the `Linker` service. The `Linker` gets all cross-references in a `LangiumDocument` and tries to resolve them. For each cross-reference, the `Linker` tries to find the correct `AstNode` and its location inside of a `LangiumDocument`. The linker uses the scopes computed during the pre-processing to identify the `AstNode` which is the target of a reference. With the default implementation, only reference that targets `AstNode`s registered in the `LangiumDocuments` service can be resolved (i.e. documents present in the current workspace). Linking can resolve references lazily if needed before the first eager resolution. 
```

Comment:
Detail: The default `ScopeProvider` uses the previously precomputed scopes, but this can be changed with a custom scope provider. The `Linker` just asks the `ScopeProvider` to create a `Scope` object for the respective context.